### ISSUE TITLE
LibARDataTransfer Samples for iOS

### ISSUE THREAD

**Issue by EthanMarsh** [ARDroneSDK3/Samples#20]
Hey guys,

Would you be willing to post some sample code for how to use libARDataTransfer with iOS to download recorded photos and videos to the device? It would be a huge help.

Thanks!

**Comment by djavan-bertrand** [comment]
Hi EthanMarsh, 

We have no sample using libARDataTransfer for the moment. But it is just a wrapper around a basic ftp.
So, if you want to use your own implementation of ARDataTransfer to download you just have to make an ftp on your device, port 21 (login anonymous). Your medias will be in internal_000/Bebop_Drone/medias.

If you want to use libARDataTransfer : 
* init ARDatatransfer : ARDATATRANSFER_Manager_New
* init 2 ftp (one for the list, the other for the download) :
  * ARUTILS_Manager_InitWifiFtp (for the Bebop)
  * ARUTILS_Manager_InitWifiFtp
* Create a mediaDownloader : ARDATATRANSFER_MediasDownloader_New
* Get the list of medias available : ARDATATRANSFER_MediasDownloader_GetAvailableMediasSync
* Get the mediaObject : ARDATATRANSFER_MediasDownloader_GetAvailableMediaAtIndex
* Download the mediaObject : ARDATATRANSFER_MediasDownloader_AddMediaToQueue

Best regards, 
Djavan

**Comment by EthanMarsh** [comment]
great thanks! I didn't realize I needed two FTP connections. I have couple other questions.

The GetAvailableMediasSync always returns 0, even though I know there are media on the Bebop (checked with FreeFlight3 to be sure). I'm passing in the mediaDownloader, and I've tried 0 and 1 for withThumbnail but I get the same result. It always returns 0 and there is no error.

If i call GetAvailableMediasSync, where can I find the media? By that I mean the thumbnails and meta data. It only returns the number of available media, so I'm not sure where they can be found.

I can't get GetAvailableMediasAsync to work either. It doesn't give an error, but the callback is never called. Is there something specific I should be passing in for void *availableMediaArg? I've tried nil, NULL, and (__bridge void *)(self).

Do I need to call ARDATATRANSFER_MediasDownloader_QueueThreadRun? If I do, when should it be called?

Do I need to register for callbacks?

 Is there any difference when trying to use a Sumo instead of the Bebop, besides the different IP address (192.168.2.1) and directory (internal_000/Jumping_Sumo/media)? I'm asking since you wrote "ARUTILS_Manager_InitWifiFTP (for the Bebop)".

**Comment by djavan-bertrand** [comment]
Hi, 
To transfer all the medias :
In the ARDATATRANSFER_MediasDownloader_New, **the remote path is only "internal_000"**.
To get the number of medias, you have to call 
int nbMedias = ARDATATRANSFER_MediasDownloader_GetAvailableMediasSync

Then you can call ARDATATRANSFER_MediasDownloader_GetAvailableMediaAtIndex from 0 to nbMedias, and add the ARDATATRANSFER_Media_t to the queue with ARDATATRANSFER_MediasDownloader_AddMediaToQueue.
after that, you can run ARDATATRANSFER_MediasDownloader_QueueThreadRun(_manager);

And yes, for the jumping sumo it will be the same code, just the ip is different.

Hope this helps, 
Best regards, 
Djavan

**Comment by djavan-bertrand** [comment]
Maybe a code extract would be more helpful : 

```objective-c
- (void)getAllMediaAsync
{
    NSString *productIP = @"192.168.42.1";
    
    eARDATATRANSFER_ERROR result = ARDATATRANSFER_OK;
    _manager = ARDATATRANSFER_Manager_New(&result);
    int mediaListCount = 0;
    
    if (result == ARDATATRANSFER_OK)
    {
        eARUTILS_ERROR ftpError = ARUTILS_OK;
        _ftpListManager = ARUTILS_Manager_New(&ftpError);
        if(ftpError == ARUTILS_OK)
        {
            _ftpQueueManager = ARUTILS_Manager_New(&ftpError);
        }
        
        if(ftpError == ARUTILS_OK)
        {
            ftpError = ARUTILS_Manager_InitWifiFtp(_ftpListManager, [productIP UTF8String], DEVICE_PORT, ARUTILS_FTP_ANONYMOUS, "");
        }
        
        if(ftpError == ARUTILS_OK)
        {
            ftpError = ARUTILS_Manager_InitWifiFtp(_ftpQueueManager, [productIP UTF8String], DEVICE_PORT, ARUTILS_FTP_ANONYMOUS, "");
        }
        
        if(ftpError != ARUTILS_OK)
        {
            result = ARDATATRANSFER_ERROR_FTP;
        }
    }
    // NO ELSE
    
    if (result == ARDATATRANSFER_OK)
    {
        NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
        NSString *path = [paths lastObject];
        
        result = ARDATATRANSFER_MediasDownloader_New(_manager, _ftpListManager, _ftpQueueManager, "internal_000", [path UTF8String]);
    }
    
    if (result == ARDATATRANSFER_OK)
    {
        mediaListCount = ARDATATRANSFER_MediasDownloader_GetAvailableMediasSync(_manager,0,&result);
        if (result == ARDATATRANSFER_OK)
        {
            for (int i = 0 ; i < mediaListCount && result == ARDATATRANSFER_OK; i++)
            {
                ARDATATRANSFER_Media_t * mediaObject_t = ARDATATRANSFER_MediasDownloader_GetAvailableMediaAtIndex(_manager, i, &result);
                if (result == ARDATATRANSFER_OK)
                {
                    result = ARDATATRANSFER_MediasDownloader_AddMediaToQueue(_manager, mediaObject_t, medias_downloader_progress_callback, (__bridge void *)(self), medias_downloader_completion_callback,(__bridge void*)self);
                }
            }
        }
    }
    
    if (result == ARDATATRANSFER_OK)
    {
        if (mediaListCount == 0)
        {
            NSLog(@"No medias");
        }
        else
        {
            ARDATATRANSFER_MediasDownloader_QueueThreadRun(_manager);
        }
        
    }
}

void medias_downloader_progress_callback(void* arg, ARDATATRANSFER_Media_t *media, float percent)
{
    NSLog(@"Media %s : %f", media->name, percent);
}

void medias_downloader_completion_callback(void* arg, ARDATATRANSFER_Media_t *media, eARDATATRANSFER_ERROR error)
{
    NSLog(@"Media %s completed : %i", media->name, error);
}
```

**Comment by EthanMarsh** [comment]
I have everything working now, thank you very much!

**Comment by CesarRN** [comment]
Hello djvan-bertrand

I have problems with your example, I obtain this error:
Undefined symbols for architecture armv7:
  "_curl_easy_reset", referenced from:
      _ARUTILS_WifiFtp_ResetOptions in libARUtils(ARUTILS_WifiFtp.o)

and others like this _curl_easy_* ....

I copy your method in the last example for the BeBoop 

How can I do to solve this error?

Thanks!

**Comment by katho** [comment]
This look very insteresting, need some help to get it working in java too...

**Comment by astrodomi** [comment]
Hola CesarRN,  same problem with _curl_etc etc
did you resolve? any help please 
Thanks

**Comment by CesarRN** [comment]
Hello astrodomi.

No, I didn´t resolve it   :/

I did it in another Mac and worked, I don´t know the reason.   

Good luck.

Regards,

**Comment by astrodomi** [comment]
well I resolved the error downloading iOS 7.40.0 libcurl and dropped in my project libcurl.a and libz.dylib but still have some problem with the this ardatatransfer code for rolling spider, changed only the ip as djvan-bertrand said but got an ftp error, should I use ...InitBLEFtp instead of InitWifiFtp?? also how to get the media ready in the pictures app??

**Comment by SteBon90** [comment]
Hi astrodomi, I have the same problem, but in Android environment.
This is the link to my issue: 
https://github.com/Parrot-Developers/ARSDKBuildUtils/issues/59
